src/app/ClusterStateCache.h

```diff
@@ -419,6 +419,35 @@ class ClusterStateCacheT : protected ReadClient::Callback
         return CHIP_NO_ERROR;
     }
 
+    /*
+     * Execute an iterator function that is called for every attribute
+     * in the cache.  The function is passed a concrete attribute path
+     * to every attribute in the cache.
+     *
+     * The iterator is expected to have this signature:
+     *      CHIP_ERROR IteratorFunc(const ConcreteAttributePath &path);
+     *
+     * Notable return values:
+     *      - If func returns an error, that will result in termination of any further iteration over attributes
+     *        and that error shall be returned back up to the original call to this function.
+     */
+    template <typename IteratorFunc>
+    CHIP_ERROR ForEachAttribute(IteratorFunc func) const
```

Comment:
It would result in cleaner code if we could use an iterator so I can loop via `for (auto path: cacheState.AllPathsIterator())` instead of using a callback-style approach.

Having callbacks often means creating labdas with some state which is more complex code and every invokation of this will create a separate implementation due to the template argument (although it can be smaller if we do not use this that often).

The PR summary does not explain _why_ we are adding this and its use...could you explain? Or if this is just helpful (it obviously is), adding tests together with the implementation is appropriate.

Reply:
I see we already have a couple of `ForEach` so the "be consistent" rule seems to imply that we want to preserve those pattern. That is unfortunate but will resolve my comment :(